core/src/main/scala/cats/data/Chain.scala

```diff
@@ -617,18 +620,38 @@ sealed abstract class Chain[+A] {
 
   /**
    * Remove duplicates. Duplicates are checked using `Order[_]` instance.
+   * 
+   * {{{
+   * scala> import cats.data.Chain
+   * scala> val chain = Chain(1, 2, 2, 3)
+   * scala> chain.distinct
+   * res0: cats.data.Chain[Int] = Chain(1, 2, 3)
+   * }}}
    */
-  def distinct[AA >: A](implicit O: Order[AA]): Chain[AA] = {
-    implicit val ord: Ordering[AA] = O.toOrdering
+  def distinct[AA >: A](implicit O: Order[AA]): Chain[AA] =
+    distinctBy(identity[AA])
 
-    var alreadyIn = TreeSet.empty[AA]
+  /**
+   * Remove duplicates by a predicate. Duplicates are checked using `Order[_]` instance.
+   * 
+   * {{{
+   * scala> import cats.data.Chain
+   * scala> val chain = Chain(1, 2, 3, 4)
+   * scala> chain.distinctBy(_ / 2)
+   * res0: cats.data.Chain[Int] = Chain(1, 2, 4)
+   * }}}
+   */
+  def distinctBy[B](f: A => B)(implicit O: Order[B]): Chain[A] = {
+    implicit val ord: Ordering[B] = O.toOrdering
 
-    foldLeft(Chain.empty[AA]) { (elementsSoFar, b) =>
-      if (alreadyIn.contains(b)) {
-        elementsSoFar
+    val alreadyIn = mutable.TreeSet.empty[B]
+
+    foldLeft(Chain.empty[A]) { (elementsSoFar, a) =>
```

Comment:
If we are using a mutable set here why not also use a var to the Chain and a while loop to avoid the closure?

Reply:
Yes agree. Fixed.